### PR TITLE
[wasm] Wasm TimeZoneInfo implementation VFS

### PR DESCRIPTION
**Pull request by kjpou1**
Right now WebAssembly does not have an implementation for TimeZoneInfo and instead throws exceptions.

This is an alternate implementation of https://github.com/mono/mono/pull/17617 that loads the zone information into the webassembly virtual file system using a pre-loaded file package.

__BCL Changes__
TimeZoneInfo implementation for WebAssembly to read from WASM VFS `/zoneinfo`.
- Remove `TimeZoneInfo.WebAssembly.cs` file
- Remove source dependencies on `TimeZoneInfo.WebAssembly.cs`.
- Modify TimeZoneInfo for specific WASM functionality
   - Default root directory is `zoneinfo`
   -  Add icall `mono_timezone_get_local_name` for WASM in CreateLocal
- Add icall implementation
   - ves_icall_System_TimeZoneInfo_mono_timezone_get_local_name

__WASM Runtime Build__

- Add `-s FORCE_FILESYSTEM=1` to mono wasm runtime build
   - Required so that it includes support for loading pre preload packages
   * Message when generating the zoneinfo data.
   > Remember to build the main file with  -s FORCE_FILESYSTEM=1  so that it includes support for loading this file package
- Add targets to run time zone tests.
   - make run-tzd-sample
   - make run-tzd-interp-sample
   - make run-tzd-aot-sample
- Modify `runtime-tests.js` to load and parse the zoneinfo data manually to load into the VFS.

__WASM Driver - aka `driver.c`__

- Add implementations of the following:
   - mono_timezone_get_local_name

These will be called from the managed bcl module `TimeZoneInfo.cs` module.

__packager.exe__
- Add `-s FORCE_FILESYSTEM=1`
   - Required so that it includes support for loading pre preload packages
   * Message when generating the zoneinfo data.
   > Remember to build the main file with  -s FORCE_FILESYSTEM=1  so that it includes support for 

__Zone Info data and support files__

- Three files for zoneinfo VFS support.
   - `mono-webassembly-zoneinfo-fs-smd.js.metadata` - This is the separate metadata output by emscripten file-packager.  Used to parse the zoneinfo data in non browser environments.  See `runtime-tests.js`
   - `mono-webassembly-zoneinfo-fs.js` - Output by emscripten file-packager that will be referenced by browser environments to load the zoneinfo data into the VFS at `/zoneinfo`
   - `zoneinfo.data` - The binary file output by emscripten file-packager.  The file contains the actual zoneinfo data loaded into the VFS and parsed by `TimeZoneInfo`.

The packages are generated automatically and the repo is temporarily at:  https://github.com/kjpou1/mono-webassembly-zoneinfo

There are some limitations to this method where the file packages generated by the emscripten [file_packager](https://emscripten.org/docs/porting/files/packaging_files.html#packaging-using-the-file-packager-tool) only runs within a web browser.

Special consideration is needed to load the VFS with the data outside of browser.  See modification for `runtime-tests.js` where the zoneinfo data is parsed manually and loaded into the VFS.  We may need to create our own file packager to handle some of the other cases.

### CODE REVIEW

**Comment by vargaz** [comment]
Why is this better than the other one ? The VFS is mostly there to support apps which already have file system access, it adds a non-trivial size overhead.

**Comment by kjpou1** [comment]
That is actually what I am trying to measure with the PR.

So far there is no size difference with the mono runtime.

```
With FORCE_FILE_SYSTEM
-rw-r--r--  1 Jimmy  staff   502263 Nov  8 13:50 ./builds/debug/mono.js
-rw-r--r--  1 Jimmy  staff  9623773 Nov  8 13:50 ./builds/debug/mono.wasm
-rw-r--r--  1 Jimmy  staff   743950 Nov  8 13:50 ./builds/debug/mono.wasm.map

Without FORCE_FILE_SYSTEM
-rw-r--r--  1 Jimmy  staff   504559 Nov  8 13:55 ./builds/debug/mono.js
-rw-r--r--  1 Jimmy  staff  9623773 Nov  8 13:55 ./builds/debug/mono.wasm
-rw-r--r--  1 Jimmy  staff   743950 Nov  8 13:55 ./builds/debug/mono.wasm.map


With FORCE_FILE_SYSTEM
-rw-r--r--  1 Jimmy  staff   210920 Nov  8 13:50 ./builds/release/mono.js
-rw-r--r--  1 Jimmy  staff  1739252 Nov  8 13:50 ./builds/release/mono.wasm

Without FORCE_FILE_SYSTEM
-rw-r--r--  1 Jimmy  staff   210138 Nov  8 13:55 ./builds/release/mono.js
-rw-r--r--  1 Jimmy  staff  1739252 Nov  8 13:55 ./builds/release/mono.wasm
``` 

Less icalls and less code added to `driver.c`.
Less interaction/roundtripping between calls of managed <-> native <-> javascript.

**Comment by kjpou1** [comment]
@monojenkins build failed

**Comment by kjpou1** [comment]
@monojenkins build failed

File: mono/metadata/icall-def.h [hunk 1]
```diff
@@ -1105,7 +1105,10 @@ NOHANDLES(ICALL(VOLATILE_6, "Write(ulong&,ulong)", ves_icall_System_Threading_Vo
 ICALL_TYPE(WAITH, "System.Threading.WaitHandle", WAITH_1)
 HANDLES(WAITH_1, "SignalAndWait_Internal", ves_icall_System_Threading_WaitHandle_SignalAndWait_Internal, gint32, 3, (gpointer, gpointer, gint32))
 HANDLES(WAITH_2, "Wait_internal", ves_icall_System_Threading_WaitHandle_Wait_internal, gint32, 4, (gpointer_ptr, gint32, MonoBoolean, gint32))
-
+#if defined(TARGET_WASM)
+ICALL_TYPE(TZINFO, "System.TimeZoneInfo", TZINFO_1)
+NOHANDLES(ICALL(TZINFO_1, "mono_timezone_get_local_name", ves_icall_System_TimeZoneInfo_mono_timezone_get_local_name))
+#endif
```
Comment (vargaz) [inline mono/metadata/icall-def.h]: Where is this defined ?

File: sdks/wasm/runtime-tests.js [hunk 2]
```diff
@@ -154,7 +158,40 @@ var Module = {
 			var f = Module.cwrap ('mono_wasm_enable_on_demand_gc', 'void', []);
 			f ();
 		}
-
+		if (enable_zoneinfo) {
```
Comment (vargaz) [inline sdks/wasm/runtime-tests.js]: This is a test file, how is this done in an app ?

File: mono/metadata/icall-def.h (hunk 1 above)
Comment (kjpou1) [reply mono/metadata/icall-def.h]: Working on adding this back after the rebase.  Should be out there soon.

File: sdks/wasm/runtime-tests.js (hunk 2 above)
Comment (kjpou1) [reply sdks/wasm/runtime-tests.js]: In the description of the PR there is a section on Zone Info data and support files.  This is loaded in the browser.  The code in the zone info data files contains code to load the zoneinfo data into the VFS.  Basically the same code as the test file. 

The package files generated do not contain code to load via node.js or code executed from JS engines.

**Comment by lewing** [comment]
@monojenkins build failed

**Comment by lewing** [comment]
This packages the all of /zoneinfo as one .data is there any reason we couldn't do it by directory?

**Comment by kjpou1** [comment]
Should not be a prp

> This packages all of /zoneinfo as one .data is there any reason we couldn't do it by directory?

Should not be a problem.  Will need to check into it.

**Comment by kjpou1** [comment]
@monojenkins build

**Comment by kjpou1** [comment]
@monojenkins build failed

File: mono/metadata/icall.c [hunk 3]
```diff
@@ -8208,6 +8208,20 @@ ves_icall_System_IO_Compression_DeflateStreamNative_WriteZStream (gpointer strea
 }
 
 #endif
+
+#if defined(TARGET_WASM)
+G_EXTERN_C void mono_timezone_get_local_name (MonoString *volatile*result);
+void
```
Comment (vargaz) [inline mono/metadata/icall.c:8214]: Is this needed ? Its already declared in icall-decl.h.

File: mono/metadata/icall.c [hunk 4]
```diff
@@ -8208,6 +8208,20 @@ ves_icall_System_IO_Compression_DeflateStreamNative_WriteZStream (gpointer strea
...
+ves_icall_System_TimeZoneInfo_mono_timezone_get_local_name (MonoString *volatile*result)
+{
```
Comment (vargaz) [inline mono/metadata/icall.c:8216]: volatile is not needed.

File: mono/metadata/icall.c [hunk 5]
```diff
@@ -8208,6 +8208,20 @@ ves_icall_System_IO_Compression_DeflateStreamNative_WriteZStream (gpointer strea
...
+	// MONO_CROSS_COMPILE returns undefined symbol "_mono_timezone_get_local_name"
+	// The icall offsets will be generated and linked at build time
+#ifndef MONO_CROSS_COMPILE
```
Comment (vargaz) [inline mono/metadata/icall.c:8219]: Should add a comment that this is defined outside the runtime.

File: sdks/wasm/src/driver.c [hunk 6]
```diff
@@ -733,3 +733,27 @@ mono_wasm_enable_on_demand_gc (void)
...
+	} catch(e) {} 
+
+	var buff = Module._malloc((res.length + 1) * 2);
+	stringToUTF16 (res, buff, (res.length + 1) * 2);
+	return buff;
+})
+
+void
+mono_timezone_get_local_name (MonoString *volatile*result)
+{
```
Comment (vargaz) [inline sdks/wasm/src/driver.c:752]: volatile is not needed.

File: sdks/wasm/Makefile [hunk 7]
```diff
@@ -497,6 +499,34 @@ build-zlib-aot-hello: packager.exe $(WASM_TESTS_BIN)/zlibtest.exe
 run-zlib-aot-hello: build-zlib-aot-hello
 	(cd bin/aot-zlib-sample && $(SM) runtime.js --run zlibtest.exe)
 
+# TimeZoneInfo tests
+
+$(TEST_SOURCE_DIR)/tzdtest.exe: $(TEST_SOURCE_DIR)/tzdtest.cs
+	$(CSC) $(CSC_FLAGS) -out:$@ $(BCL_DEPS) $(TEST_SOURCE_DIR)/tzdtest.cs
+
+run-tzd-sample: build-tzd-sample
+	(cd bin/tzdtest-sample && $(SM) runtime.js --enable-zoneinfo --run tzdtest.exe)
```
Comment (vargaz) [inline sdks/wasm/Makefile:508]: Is this a sample or a test ?

Comment (kjpou1) [reply sdks/wasm/Makefile:508]: I will address this in a separate PR.

**Comment by kjpou1** [comment]
@monojenkins build Linux WebAssembly

**Comment by kjpou1** [comment]
@monojenkins build OS X x64 iOS SDK

**Comment by kjpou1** [comment]
@monojenkins build Linux WebAssembly

**Comment by lewing** [comment]
@monojenkins build failed

**Pull request merged**